### PR TITLE
Filter registration

### PR DESCRIPTION
**Pull request by ethomson**
While reviewing #3564 , @carlosmn noticed that we were a bit racy in registration of merge drivers.  This registration pattern was primarily copied from filter registration, so I wanted to fix up the existing code before banging on that.

Previously we would set the global filter registry structure before adding filters to the structure, without a lock, which is quite racy. Now, register default filters during global registration and use an rwlock to read and write the filter registry (as appropriate)

### CODE REVIEW

**Comment by ethomson** [comment]
Okay - I nailed down the problems on Windows; it turns out that my refactoring was a bit overly aggressive when it came to the windows CRT debug functions.  Those *must* be first, before your first allocation.  This is quite sensible, but not something that had occurred to me.

In any case, I think this nicely tightens up the global setup and teardown, and makes our threads safer for the filters.

This should be ready for review.

File: src/openssl_stream.c [hunk 1]
```diff
@@ -31,6 +31,115 @@
...
+		return -1;
+	}
+#endif
+
+	git__on_shutdown(shutdown_ssl);
+
+	return 0;
+}
+
+int git_openssl_set_locking(void)
```
Comment (carlosmn) [inline src/openssl_stream.c]: Thing is... this function isn't really about the streams. It's there to stop bindings from crashing when they use multiple threads without having to conditionally link to OpenSSL themselves.

Comment (ethomson) [inline src/openssl_stream.c]: Yes but short of deleting it, t this still seems to be the most logical file to move this function to.

**Comment by stanhu** [comment]
Does a similar change have to be made with the diff driver registry?

**Comment by ethomson** [comment]
What diff driver registry?

**Comment by stanhu** [comment]
I may be mistaken, but I thought this was it:

https://github.com/libgit2/libgit2/blob/master/src/diff_driver.h#L13

**Comment by ethomson** [comment]
Ah - no, those are per-repository and so have very different lifecycles.

**Pull request merged**